homeassistant/components/alexa/smart_home.py

```diff
@@ -0,0 +1,185 @@
+"""Support for alexa Smart Home Skill API."""
+import asyncio
+import logging
+from uuid import uuid4
+
+from homeassistant.const import (
+    ATTR_SUPPORTED_FEATURES, ATTR_ENTITY_ID, SERVICE_TURN_ON, SERVICE_TURN_OFF)
+from homeassistant.components import switch, light
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_HEADER = 'header'
+ATTR_NAME = 'name'
+ATTR_NAMESPACE = 'namespace'
+ATTR_MESSAGE_ID = 'messageId'
+ATTR_PAYLOAD = 'payload'
+ATTR_PAYLOAD_VERSION = 'payloadVersion'
+
+
+MAPPING_COMPONENT = {
+    switch.DOMAIN: ['SWITCH', ('turnOff', 'turnOn'), None],
+    light.DOMAIN: [
+        'LIGHT', ('turnOff', 'turnOn'), {
+            light.SUPPORT_BRIGHTNESS: 'setPercentage'
+        }
+    ],
+}
+
+
+def mapping_api_function(name):
+    """Return function pointer to api function for name.
+
+    Async friendly.
+    """
+    mapping = {
+        'DiscoverAppliancesRequest': async_api_discovery,
+        'TurnOnRequest': async_api_turn_on,
+        'TurnOffRequest': async_api_turn_off,
+        'SetPercentageRequest': async_api_set_percentage,
+    }
+    return mapping.get(name, None)
+
+
+@asyncio.coroutine
+def async_handle_message(hass, message):
+    """Handle incomming API messages."""
+    assert int(message[ATTR_HEADER][ATTR_PAYLOAD_VERSION]) == 2
+
+    # Do we support this API request?
+    funct_ref = mapping_api_function(message[ATTR_HEADER][ATTR_NAME])
+    if not funct_ref:
+        _LOGGER.warning(
+            "Unsupported API request %s", message[ATTR_HEADER][ATTR_NAME])
+        return api_error(message)
+
+    return (yield from funct_ref(hass, message))
+
+
+def api_message(name, namespace, payload=None):
+    """Create a API formated response message.
+
+    Async friendly.
+    """
+    payload = payload or {}
+    return {
+        ATTR_HEADER: {
+            ATTR_MESSAGE_ID: uuid4(),
+            ATTR_NAME: name,
+            ATTR_NAMESPACE: namespace,
+            ATTR_PAYLOAD_VERSION: '2',
+        },
+        ATTR_PAYLOAD: payload,
+    }
+
+
+def api_error(request, exc='DriverInternalError'):
+    """Create a API formated error response.
+
+    Async friendly.
+    """
+    return api_message(exc, request[ATTR_HEADER][ATTR_NAMESPACE])
+
+
+@asyncio.coroutine
+def async_api_discovery(hass, request):
+    """Create a API formated discovery response.
+
+    Async friendly.
+    """
+    discovered_appliances = []
+
+    for domain, class_data in MAPPING_COMPONENT.items():
```

Comment:
I think that it would be faster if we turn the loops around:

```python
for entity in hass.states.async_all():
    class_data = MAPPING_COMPONENT.get(entity.domain)

    if class_data is None:
        continue

    …
```

Reply:
True